### PR TITLE
Update check for classic vs. block theme in the WooCommerce Analytics module

### PR DESCRIPTION
**Pull request by opr**
## Proposed changes:

* Adds a check to determine if a classic theme is in use. That way we will only query the Cart/Checkout _pages_ and not the _templates_ when determining block vs. shortcode use and additional blocks used in the WooCommerce Analytics module.
* Updates the type of the `cart_page_contains_cart_block`, `cart_page_contains_cart_shortcode`, `checkout_page_contains_checkout_block`, and `checkout_page_contains_checkout_shortcode` properties so it is always a `0` or `1` rather than `true` or `false`.
* Checks for the presence of the `woocommerce/classic-shortcode` block as well as the normal shortcode
* Updates the check in `class-jetpack-woocommerce-analytics-universal.php` when getting additional blocks on a template to check for the template by its other possible name, then if still not found it bails cleanly without generating a warning.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

N/A

## Does this pull request change what data or activity we track or use?

No, it just updates the accuracy of the properties.

## Testing instructions:

*Please note, the properties for the Cart/Checkout Shortcode/Blocks use is fetched only once per day, and is stored in a transient, you will need to clear the `jetpack_woocommerce_analytics_cart_checkout_content_sources`, `jetpack_woocommerce_analytics_additional_blocks_on_checkout_page`, and `jetpack_woocommerce_analytics_additional_blocks_on_cart_page` transients while testing to ensure the most up to date data*

**Please set up a fresh site before testing**

Throughout testing, please check for PHP warnings. There should not be any.

### Classic theme testing (e.g. Storefront)

- Using a classic theme (e.g. Storefront) edit your Cart and Checkout pages, set them to be using the Cart and Checkout _blocks_.
- Add items to your cart and go to the Cart page.
- Using tracks vigilante (p7H4VZ-3cB-p2), ensure the following properties on the `woocommerceanalytics_cart_view` event contain:
  - `cart_page_contains_cart_block`: `1`
  - `cart_page_contains_cart_shortcode`: `0`
  - `checkout_page_contains_checkout_block`: `1`
  - `checkout_page_contains_checkout_shortcode`: `0`
- Proceed to the Checkout block and check the same properties and their values on the `woocommerceanalytics_product_checkout` and `woocommerceanalytics_checkout_view` events.
- Change your Cart/Checkout pages to use the shortcode block (enter `[woocommerce_cart]` on the Cart page, and `[woocommerce_checkout]` on the Checkout page).
- Proceed to the Checkout page again (note the `woocommerceanalytics_checkout_view` will not happen in a shortcode checkout, so don't expect it):
  - `cart_page_contains_cart_block`: `0`
  - `cart_page_contains_cart_shortcode`: `1`
  - `checkout_page_contains_checkout_block`: `0`
  - `checkout_page_contains_checkout_shortcode`: `1`

### Block theme testing (e.g. Twenty Twenty-Three)

- Switch to a block theme and repeat the above steps, however edit the Cart/Checkout templates rather than the pages.
- Install an incompatible extension (e.g. Checkout Field Editor).
- Delete *all* blocks from the pages before saving each time!
- Go to the Checkout template and select the Checkout block. In the block sidebar it will ask you if you'd like to revert to the classic Checkout.
image

- Click that, and wait for the classic shortcode block to show, it looks like this:
image

- Save your changes and do the same on the Cart template.
- Go to the Checkout block and ensure the properties on the `woocommerceanalytics_product_checkout` are:
  - `cart_page_contains_cart_block`: `0`
  - `cart_page_contains_cart_shortcode`: `1`
  - `checkout_page_contains_checkout_block`: `0`
  - `checkout_page_contains_checkout_shortcode`: `1`

### Page content on block themes

- Go to Appearance -> Editor -> Templates -> Page: Cart
- Click the three dots at the top and choose "Clear customizations"

image

- If the dots aren't there, make some changes to the template, save it, then close the tab and reopen the editor in a new tab.

- Go to Appearance -> Editor -> Templates -> Page: Checkout
- Click the three dots at the top and choose "Clear customizations" like above.

Check that both templates are using the "Page content" block. (If you edit the template and open the list view, it should look like this:)
image

- Go back to the Admin dashboard front page. Repeat the testing steps from **Classic theme testing (e.g. Storefront)**

### CODE REVIEW

**Review by jeherve — DISMISSED** [review]

**Comment by opr** [comment]
@jeherve thanks for your review, I updated the code to cater for the "classic shortcode" block. I updated the testing instructions to reflect this. @senadir Let me know if you need any help testing this.

I am also getting E2E fails relating to Mailchimp connection, I see other PRs are failing with this too, so wondering whether we can merge with it failing?

**Review by jeherve — DISMISSED** [review]

**Comment by opr** [comment]
@senadir please can you review this when you get chance? 🙏🏼

**Review by senadir — CHANGES_REQUESTED** [review]
As discussed in our call, this code doesn't yet support the new `woocommerce/page-content-wrapper` setup we have and that's why it didn't work. Do request another review once that fix is done.

File: projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-trait.php
Comment (senadir) [inline projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-trait.php:160]: It would help greatly if we have a list here of the steps we go through to get the logic as I'm finding it hard to follow if we already achieved our goal, branched out to something else, or still trying to get the content.

File: projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php
Comment (senadir) [inline projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php]: A lot of those seems to be shared between events, can we move them all to a single place so we don't make mistakes?

File: projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-trait.php
Comment (nielslange) [inline projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-trait.php]: Nitpic: I wonder if we can increase the readability by avoiding breaks here. The preview looks a bit strange, though. This is how the suggestion looks in the editor:

Screenshot 2023-12-04 at 21 21 34

```suggestion
public function get_cart_page_block_usage() {
	$new_info = array();

	$content                    = $this->cart_content_source;
	$block_presence             = str_contains( $content, '<!-- wp:woocommerce/cart' );
	$shortcode_presence         = str_contains( $content, '[woocommerce_cart]' );
	$classic_shortcode_presence = str_contains( $content, '<!-- wp:woocommerce/classic-shortcode' );

	$new_info['cart_page_contains_cart_block']     = $block_presence ? '1' : '0';
	$new_info['cart_page_contains_cart_shortcode'] = $shortcode_presence || $classic_shortcode_presence ? '1' : '0';

	return $new_info;
}
```

File: projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-trait.php
Comment (nielslange) [inline projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-trait.php]: Nitpic: Similar to my previous suggestion, I think we can also improve the readability in the following part. This is how the suggested change looks in the editor:

Screenshot 2023-12-04 at 21 24 18

```suggestion
public function get_checkout_page_block_usage() {
	$new_info = array();

	$content                    = $this->checkout_content_source;
	$block_presence             = str_contains( $content, '<!-- wp:woocommerce/checkout' );
	$shortcode_presence         = str_contains( $content, '[woocommerce_checkout]' );
	$classic_shortcode_presence = str_contains( $content, '<!-- wp:woocommerce/classic-shortcode' );

	$new_info['checkout_page_contains_checkout_block']     = $block_presence ? '1' : '0';
	$new_info['checkout_page_contains_checkout_shortcode'] = $shortcode_presence || $classic_shortcode_presence ? '1' : '0';

	return $new_info;
}
```

File: projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-trait.php
Comment (nielslange) [inline projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-trait.php]: Nitpic: I'd add a line break here to increase the readability.

```suggestion
			if ( $cart_page && isset( $cart_page->post_content ) ) {
				$this->cart_content_source = $cart_page->post_content;
			}
			
			if ( $checkout_page && isset( $checkout_page->post_content ) ) {
				$this->checkout_content_source = $checkout_page->post_content;
			}
```

File: projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-trait.php
Comment (nielslange) [inline projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-trait.php]: Nicpic: Adding a few more line breaks due to readability.

```suggestion
		if ( function_exists( 'get_block_template' ) ) {
			$checkout_template = get_block_template( 'woocommerce/woocommerce//page-checkout' );
			$cart_template     = get_block_template( 'woocommerce/woocommerce//page-cart' );
			
			if ( ! $checkout_template ) {
				$checkout_template = get_block_template( 'woocommerce/woocommerce//checkout' );
			}
			
			if ( ! $cart_template ) {
				$cart_template = get_block_template( 'woocommerce/woocommerce//cart' );
			}
		}
```

File: projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-trait.php
Comment (nielslange) [inline projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-trait.php]: Nitpic: _"Good things come in threes."_ Suggesting a few more line breaks to increase the readability.

```suggestion
		if ( function_exists( 'gutenberg_get_block_template' ) ) {
			$checkout_template = gutenberg_get_block_template( 'woocommerce/woocommerce//page-checkout' );
			$cart_template     = gutenberg_get_block_template( 'woocommerce/woocommerce//page-cart' );
			
			if ( ! $checkout_template ) {
				$checkout_template = gutenberg_get_block_template( 'woocommerce/woocommerce//checkout' );
			}
			
			if ( ! $cart_template ) {
				$cart_template = gutenberg_get_block_template( 'woocommerce/woocommerce//cart' );
			}
		}
```

File: projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-trait.php
Comment (nielslange) [inline projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-trait.php]: Nitpic: As we're only using $cart_page_id and $checkout_page_id to retrieve the post, I wonder if we should simplify this part as follows:

```suggestion
			$cart_page        = get_post( wc_get_page_id( 'cart' ) );
			$checkout_page    = get_post( wc_get_page_id( 'checkout' ) );
```

File: projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-trait.php
Comment (nielslange) [inline projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-trait.php]: Nitpic: Similar to my other suggestions, I wonder if we should simplify this part as well:

```suggestion
				$checkout_page    = get_post( wc_get_page_id( 'checkout' ) );
```

File: projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-trait.php
Comment (nielslange) [inline projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-trait.php]: Nicpic: If we'd simplify the other parts, I previously mentioned, we could also simplify this part here:

```suggestion
				$cart_page    = get_post( wc_get_page_id( 'cart' ) );
```

File: projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php
Comment (nielslange) [inline projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php]: I wonder what the benefit of using `Y` and `N` over `Yes` and `No` are? If we'd use `Yes` and `No`, we could simply use the following part:

```suggestion
						'guest_checkout'   => $guest_checkout,
						'create_account'   => '$create_account,
```

If we'd proceed with `Y` and `N`, we might want to adjust the corresponding part in `woocommerceanalytics_product_purchase` accordingly.

File: projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php
Comment (nielslange) [inline projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php]: See my previous comment. I wonder why we don't use the following part:

```suggestion
						'guest_checkout'   => $guest_checkout,
						'create_account'   => $create_account,
```

File: projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php
Comment (nielslange) [inline projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php]: Nitpic: Similar as before. I wonder if we should add some line breaks to increase the readability. There's also an existing part below this function, which cannot be reviewed/commented, but could benefit from adding a line break.

```suggestion
				if ( 'woocommerce/classic-shortcode' === $block['blockName'] ) {
					return false;
				}
				
				if ( 'core/shortcode' === $block['blockName'] ) {
					return false;
				}
				
				if ( 'checkout' === $cart_or_checkout && 'woocommerce/checkout' !== $block['blockName'] ) {
```

**Review by senadir — DISMISSED** [review]
I'm pre-approving this PR because we addressed everything that needs addressing.

**Review by nielslange — COMMENTED** [review]
Thanks for working on this, @opr. In my first round of review, mainly focused on readability. My suggestions are nitpics. I'll further review this PR, but wanted to share this feedback already with you.

File: projects/plugins/jetpack/modules/woocommerce-analytics/class-jetpack-woocommerce-analytics.php
Comment (senadir) [inline projects/plugins/jetpack/modules/woocommerce-analytics/class-jetpack-woocommerce-analytics.php]: Forget to mention, this should be feature gated with 11.6.2

**Review by nielslange — DISMISSED** [review]
Thanks for working on this, @opr. I've successfully tested the PR based on the provided testing steps. Just one minor note, you mentioned _"clear the `jetpack_woocommerce_analytics_cart_checkout_info_cache` transient"_. On my end, that transient is called `jetpack_woocommerce_analytics_cart_checkout_content_sources`.

**Comment by opr** [comment]
@senadir @nielslange - thanks for your reviews. I just pushed [8834b8a](https://github.com/Automattic/jetpack/pull/34065/commits/8834b8a942977725d5a4d5baac6c75a946f12a2e) which adds a check for WC Blocks version - please re-review and check the page view events only happen once, then test with WC Blocks < 11.6.2 (install from an old zip and use an older version of WC) and ensure the page view events still fire only once.

**Comment by senadir** [comment]
I might not be able to test this today @nielslange if you get a chance can you test it and get it merged?

**Review by nielslange — COMMENTED** [review]
I've successfully tested this PR, @opr. I do have a question regarding this comment from you, though:

> [...] please re-review and check the page view events only happen once, then test with WC Blocks < 11.6.2 (install from an old zip and use an older version of WC) and ensure the page view events still fire only once.

On the cart page, I only see the `woocommerceanalytics_cart_view` event. However, on the checkout page, I see the `woocommerceanalytics_checkout_view` and the `woocommerceanalytics_product_checkout` events. Is that expected?

**Comment by opr** [comment]
> I've successfully tested this PR, @opr. I do have a question regarding this comment from you, though:
> ...
> On the cart page, I only see the `woocommerceanalytics_cart_view` event. However, on the checkout page, I see the `woocommerceanalytics_checkout_view` and the `woocommerceanalytics_product_checkout` events. Is that expected?

Yes that's correct!

**Comment by opr** [comment]
@jeherve hi! Thanks for your patience while we get this merged. It's been thoroughly tested by our team and we're just waiting for @nielslange to approve.

Could you also review and approve the pr when able?

Thanks!

**Review by nielslange — APPROVED** [review]
Thanks for answering the question I had during my most recent review, @opr. As you haven't applied any changes, and I've successfully tested the PR, I'm approving it now. Thanks again for working on this. 🙌

File: projects/plugins/jetpack/modules/woocommerce-analytics/class-jetpack-woocommerce-analytics.php
Comment (jeherve) [inline projects/plugins/jetpack/modules/woocommerce-analytics/class-jetpack-woocommerce-analytics.php:111]: Could this be added via `wp_add_inline_script()`, hooked into the Tracks script, instead?

File: projects/plugins/jetpack/modules/woocommerce-analytics/class-jetpack-woocommerce-analytics.php
Comment (jeherve) [inline projects/plugins/jetpack/modules/woocommerce-analytics/class-jetpack-woocommerce-analytics.php:122]: You could use `register_script` here instead, since there is no difference between minimized and full script.

**Review by jeherve — COMMENTED** [review]
I'm afraid I'm not familiar enough with Woo Analytics to fully test this, but this looks good to me. I left a few minor notes, below.

> Does this pull request change what data or activity we track or use?
> No, it just updates the accuracy of the properties.

This would need to be updated, and documented in our privacy policy settings. We're now loading a stats script from a third-party (wp.com) on every page of the site, in a module that's automatically activated when the WooCommerce plugin is active.

**Comment by opr** [comment]
@jeherve  thank you for the review - your inline code comments relate to "old" code that was not added as part of this PR. It was moved to a different file as-is.

The script from wp.com that you mentioned (re the privacy notice) was already here too, just moved files in this PR.

I appreciate you pointing these out but I think making changes here would expand the scope of the PR beyond its original intention. I think the Woo Analytics module in general could do with a larger clean-up so perhaps better to log this and tackle these things during that effort, do you agree?

**Comment by jeherve** [comment]
> your inline code comments relate to "old" code that was not added as part of this PR. It was moved to a different file as-is.
> 
> The script from wp.com that you mentioned (re the privacy notice) was already here too, just moved files in this PR.
> 
> I appreciate you pointing these out but I think making changes here would expand the scope of the PR beyond its original intention. I think the Woo Analytics module in general could do with a larger clean-up so perhaps better to log this and tackle these things during that effort, do you agree?

ah right, my bad!

I agree, yep!

**Pull request merged**
